[Calendar] Meeting Room Booking

xhr0804 opened the pull request:
<!--- This repository only accepts pull requests related to open issues, please link the open issue in description below. See https://help.github.com/articles/closing-issues-using-keywords/ to learn about automation. 
For example - Close #123: Description goes here. -->
### Purpose
*What is the context of this pull request? Why is it being done?*

close #934 

### Changes
*Are there any changes that need to be called out as significant or particularly difficult to grasp? (Include illustrative screenshots for context if applicable.)*

implement BookMeetingRoom/CancelMeetingRoom/AddMeeetingRoom/ChangeMeetingRoom and add room feature for the CreateMeeting

**Book a meeting room:**

![image](https://user-images.githubusercontent.com/44257569/71144436-f6204d80-2258-11ea-95a4-449ca4dfc14b.png)
![image](https://user-images.githubusercontent.com/44257569/71144441-f7ea1100-2258-11ea-8056-e3f7a168bcc7.png)

**Cancel meeting room:**

![image](https://user-images.githubusercontent.com/44257569/71144447-020c0f80-2259-11ea-8e9f-3bc087fca4c4.png)

**Add meeting room:**

![image](https://user-images.githubusercontent.com/44257569/71144467-105a2b80-2259-11ea-851c-dddee17b30da.png)

**Change meeting room:**

![image](https://user-images.githubusercontent.com/44257569/71144473-18b26680-2259-11ea-9da6-ca76d39df13d.png)

### Tests
*Is this covered by existing tests or new ones? If no, why not?*

### Feature Plan
*Are there any remaining steps or dependencies before this issue can be fully resolved? If so, describe and link to any relevant pull requests or issues.*
Will provide a tool to import data into cosmosDb/AzureSearch from msgraph.

### Checklist

#### General
- [ ] I have commented my code, particularly in hard-to-understand areas	
- [ ] I have added or updated the appropriate tests	
- [ ] I have updated related documentation

#### Bots
- [ ] I have validated that new and updated responses use appropriate [Speak](https://docs.microsoft.com/en-us/azure/bot-service/dotnet/bot-builder-dotnet-text-to-speech?view=azure-bot-service-3.0) and [InputHint](https://docs.microsoft.com/en-us/azure/bot-service/dotnet/bot-builder-dotnet-add-input-hints?view=azure-bot-service-3.0) properties to ensure a high-quality speech-first experience
- [ ] I have replicated language model changes across the English, French, Italian, German, Spanish, and Chinese `.lu` files and validated that deployment is successful

#### Deployment Scripts
- [ ] I have replicated my changes in the **Virtual Assistant Template** and **Sample** projects
- [ ] I have replicated my changes in the **Skill Template** and **Sample** projects

darrenj commented on skills/csharp/calendarskill/Deployment/Resources/LU/en-us/Calendar.lu:
```diff
@@ -32,6 +32,49 @@
 
 
+## AddAttendee
+- add {ContactName=carlos} to the meeting
```
Any reason for not using personname pre-built entity? Are these changes run past the LUIS pre-built team?

xhr0804 replied:
This intent and the utterances are provided by LUIS team. According to LUIS team, the reason for not using pre-built personName entity is that the personName has some limitation. For example, we can not use multiple personNames in one utterance. They will use personName when these limitations are resolved.

darrenj commented on skills/csharp/calendarskill/Deployment/Resources/LU/en-us/Calendar.lu:
```diff
@@ -32,6 +32,49 @@
+
+
+## AddMeetingRoom
+- add a room to my {ToTime=3pm} meeting
```
Assuming that ToTime/etc. entities are those used by other intents in this LUIS model?

xhr0804 replied:
Yes, we just reuse the ToTime/etc. entities as the other intents like "ChangeCalendarEntry"

xhr0804 replied:
LUIS team already gave us the model for several intents (FindMeetingRoom/CheckAvailability), but not all intents of the flow such as Add(here)/Change/Cancel room. 
They are still working on those intents. We will try to update when it is ready.

darrenj commented on skills/csharp/calendarskill/Deployment/Resources/LU/en-us/Calendar.lu:
```diff
@@ -195,22 +253,74 @@
 - am i free {FromDate=tuesday} at {FromTime=7 am}
 - can you check my dad 's calendar for free time {FromDate=this week}
 - can you tell me if my husband is free on {FromDate=sunday}
+- can i reserve a room in building {Building=5} for a meeting?
```
As previous comment, have these utterances been reviewed with LUIS team? Just wondering if there are enough variants of meeting room, floor number ?

xhr0804 replied:
These utterances are provided by LUIS team, will check with them for the variants.

xhr0804 replied:
We have checked the MeetingRoom variants with LUIS team. They have provided more utterances with more variants and some patterns  to recognize meeting room. We have updated this PR with the new luis model.

darrenj commented on skills/csharp/calendarskill/Deployment/Resources/LU/en-us/Calendar.lu:
```diff
@@ -224,9 +334,9 @@
-- ^do i have time (at|on|for) {FromTime} [on] {FromDate}
-- ^does {personName} have time at {FromTime}
 - ^is {personName} (available|free) from {FromTime} to {ToTime} [on {FromDate}]
+- ^does {personName} have time at {FromTime}
```
This will have to trigger checking a persons time rather than a room?

xhr0804 replied:
The intent "CheckAvailability" includes both MeetingRoom and Person, but we will use the Entity to determine which dialog will be triggered.

darrenj commented on skills/csharp/calendarskill/Deployment/Resources/LU/en-us/Calendar.lu:
```diff
@@ -1223,6 +1435,28 @@ $PREBUILT:ordinal
 
 > # List entities
 
+$BuildingName:StudioX=
```
Given these are Microsoft only - do we not need more generic training? We should seek guidance from LUIS team?

xhr0804 replied:
removed these internal entities.

darrenj commented on skills/csharp/calendarskill/Dialogs/BookMeetingRoomDialog.cs:
```diff
@@ -0,0 +1,100 @@
+            try
+            {
+                var state = await Accessor.GetAsync(sc.Context);
+                DateTime dateNow = TimeConverter.ConvertUtcToUserTime(DateTime.UtcNow, state.GetUserTimeZone());
```
only initialise when needed within the `if` block below?

xhr0804 replied:
Actually the BookMeetingRoomDialog  just has two steps, first step call FindMeetingRoom, second call the CreateEventDialog.
If we use FindMeetingRoom directly, there will be a confict. Because due to Em's design, CreateEventDialog  will call FindMeetingRoom as a subfloow, while FindMeetingRoom will also call CreateEventDialog as a subflow, this is a conflict. 
So I just create a new Dialog BookMeetingRoom, so that both BookMeetingRoom and CreateEventDialog can call the FindMeetingRoom. This can simplfy the code.

Do you think we should remove the BookMeetingRoomDialog, and merge it into one Dialog?

xhr0804 replied:
Without info about Time in user's query, we will use "right now" here to initialise.
Or we will collect all the time info in next step FindMeetingRoomDialog.

darrenj commented on skills/csharp/calendarskill/Dialogs/BookMeetingRoomDialog.cs:
```diff
@@ -0,0 +1,100 @@
+            try
+            {
+                var state = await Accessor.GetAsync(sc.Context);
+                if (state.MeetingInfor.MeetingRoom == null)
```
`MeegingInfo` a typo>

Error handling needed, if meetingroom is a property we always expect to be passed? Otherwise dialog just terminates?

xhr0804 replied:
add error handling, when meetingroom==null, throw exception

darrenj commented on skills/csharp/calendarskill/Dialogs/MainDialog.cs:
```diff
@@ -155,7 +174,14 @@ protected override async Task OnMessageActivityAsync(DialogContext dc, Cancellat
+                        }
+                        else
+                        {
+                            await dc.BeginDialogAsync(nameof(CheckAvailableDialog));
```
CheckPersonAvailableDialog ?

darrenj commented on skills/csharp/calendarskill/Dialogs/UpdateMeetingRoomDialog.cs:
```diff
@@ -0,0 +1,213 @@
+
+                if (state.InitialIntent == CalendarLuis.Intent.ChangeMeetingRoom || state.InitialIntent == CalendarLuis.Intent.AddMeetingRoom)
+                {
+                    if (state.MeetingInfor.MeetingRoom == null)
```
As before, not sure why we end the dialog in this situation - exception handling/logging needed. Validation earlier up the stack to avoid? e.g. Guard statements?

xhr0804 replied:
added error handling to handle it.

darrenj commented on skills/csharp/calendarskill/Dialogs/UpdateMeetingRoomDialog.cs:
```diff
@@ -0,0 +1,213 @@
+
+                var data = new
+                {
+                    MeetingRoom = meetingRoom ?? "",
```
Is it valid for meeting room to be null, if not we should have guard statements above - also if you really need it either set to null or `string.Empty`

xhr0804 replied:
fixed

darrenj commented on skills/csharp/calendarskill/Models/CalendarSkillState.cs:
```diff
@@ -122,6 +122,22 @@ public class MeetingInformation
+
+            public int? FloorNumber { get; set; }
+
+            public bool Allday { get; set; } = false;
```
AllDay ? (casing)

xhr0804 replied:
fixed

darrenj commented on skills/csharp/calendarskill/Models/DialogOptions/CollectInfoDialogOptions.cs:
```diff
@@ -0,0 +1,35 @@
+        public enum UpdateReason
+        {
+            /// <summary>
+            /// FirstCollect.
```
Better description to explain what it means?

darrenj commented on skills/csharp/calendarskill/Models/DialogOptions/CollectInfoDialogOptions.cs:
```diff
@@ -0,0 +1,35 @@
+    {
+        public CollectInfoDialogOptions()
+        {
+            Reason = UpdateReason.FirstCollect;
```
Perhaps simplify with 
```
  public UpdateReason Reason { get; set; } = UpdateReason.FirstCollect;
```

xhr0804 replied:
fixed, abandon this usage

darrenj commented on skills/csharp/calendarskill/Models/DialogOptions/CollectInfoDialogOptions.cs:
```diff
@@ -0,0 +1,35 @@
+
+namespace CalendarSkill.Models.DialogOptions
+{
+    public class CollectInfoDialogOptions
```
General comment - not clear what these classes are for so need simple class documentation across the board

xhr0804 replied:
abandon this option

darrenj commented on skills/csharp/calendarskill/Prompts/GetRecreateInfoPrompt.cs:
```diff
@@ -136,6 +136,14 @@ protected override async Task OnPromptAsync(ITurnContext turnContext, IDictionar
             }
 
+            // check is change meeting room
+            regex = new Regex(CalendarCommonStrings.AdjustMeetingRoom);
```
Any reason for using regex vs LUIS - is this flexible enough for the user?

xhr0804 replied:
changed to LUIS, I used regex just because it was used in some other part and the previous LUIS model can't support it well.

darrenj commented on skills/csharp/calendarskill/Responses/FindMeetingRoom/FindMeetingRoomDialogActivities.es-es.lg:
```diff
@@ -0,0 +1,148 @@
+
+# NoBuilding
+[Activity
+    Text = {NoBuilding.Text()}
```
Can you please test with the latest 4.7 LG library - this syntax has been deprecated in the latest build and needs to be prefixed with @, for example: `@{NoBuilding.Text()}

xhr0804 replied:
changed to 4.7

darrenj commented on skills/csharp/calendarskill/Services/AzureSearchAPI/AzureSearchClient.cs:
```diff
@@ -0,0 +1,64 @@
+                SearchParameters parameters = new SearchParameters()
+                {
+                    SearchMode = SearchMode.All,
+                    Filter = floorNumber == 0 ? null : "FloorNumber eq " + floorNumber.ToString()
```
FloorNumber should not be a string literal but a common, shared variable somewhere?

darrenj commented on skills/csharp/calendarskill/Services/AzureSearchAPI/AzureSearchClient.cs:
```diff
@@ -0,0 +1,64 @@
+                        DisplayName = item.Document["DisplayName"].ToString(),
+                        EmailAddress = item.Document["EmailAddress"].ToString(),
+                        Building = item.Document["Building"].ToString(),
+                        FloorNumber = int.Parse(item.Document["FloorNumber"].ToString())
```
We should make this more defensive - what if floor number isn't found in the collection because data isn't available?

darrenj replied:
Some unit tests will help with these types of issues?

xhr0804 replied:
fixed this problem. We will just treat the EmailAddress of meeting room as the required field, while others optional.

darrenj commented on skills/csharp/calendarskill/Services/MSGraphAPI/MSGraphCalendarAPI.cs:
```diff
@@ -142,6 +142,44 @@ public async Task AcceptEventByIdAsync(string id)
+
+                foreach (var page in collectionPage)
+                {
+                    availability.Add(page.AvailabilityView[0] == '0');
```
Need comments to explain what this is doing? What is there isn't an element in AvailabilityView? (error handling)

xhr0804 replied:
added comments and error handling

darrenj commented on skills/csharp/calendarskill/Utilities/Actions.cs:
```diff
@@ -64,5 +64,16 @@ public static class Actions
+        public const string CheckAvailability = "checkAvailability";
+        public const string BookConfirmedMeetingRoom = "bookConfirmedMeetingRoom";
+        public const string CollectMeetingRoom = "collectMeetingRoom";
+        public const string ConfirmReFindSpecificMeeingRoom = "confirmReFindSpecificMeeingRoom";
```
Typo? ReFind? MeeingRoom?

xhr0804 commented:
@darrenj Hi Darren, could you please review the PR again? I've fixed the above comments.

LUIS team has provided more utterences and variants for the FindMeetingRoom, we have added them into our model to improve the recognition accuracy.

I have added a script to support the FindMeetingRoom features, which will create an azure search instance and help populate cosmosdb with the meeting rooms data from MSGraph(including a sign-in step).  User may run this script if they want to use the feature.

Documents are also updated to reflect the new scenarios and the guide to use FindMeetingRoom.

bobokids commented on docs/_docs/skills/samples/calendar.md:
```diff
@@ -124,6 +134,17 @@ LUIS models for the Skill are provided in **.lu** file format as part of the Ski
 {:.no_toc}
 Learn how to [provision your Azure resources]({{site.baseurl}}/skills/tutorials/create-skill/csharp/4-provision-your-azure-resources/) in the Create a Skill tutorial.
 
+To use the "FindMeetingRoom" feature, you need follow these steps:
```
How about we move this part as a standalone section after ###Authentication connection settings?

### Support of search and book a meeting room
Calendar skill provided additional support to search and book a meeting room in the conversation. You can get the meeting room data from the msgraph API and deploy the data to Azure Search Index to provide better search experience.
1. Make sure your meeting room data is complaint with the schema
<Json schema>
2. Configure the setting of your registered app in Azure App Registration portal
        i. Make sure your account have the admin privileges to access your tenant's meeting room data. 
        ii. In API Permissions, add Scopes: **Place.Read.All**
3. Run the following command 
        "./Deployment/Scripts/Enable_findmeetingroom.ps1"
        ### What do these parameters mean?
       <table for parameters>
       When running the script, you will be asked to sign in with your account which can access the meeting room data in the MSGraph.

xhr0804 replied:
sure, updated

bobokids commented on docs/_docs/skills/samples/calendar.md:
```diff
@@ -97,10 +101,12 @@ LUIS models for the Skill are provided in **.lu** file format as part of the Ski
 |Name|Description|
 |-|-|
 |AskParameter| Simple entity|
+|building| Simple entity|
```
Captical character

xhr0804 replied:
fixed

bobokids commented on docs/_docs/skills/samples/calendar.md:
```diff
@@ -47,6 +47,9 @@ The following scenarios are currently supported by the Skill:
 - Find an Event's Duration
   - *How long will the next meeting last?*
   - *What's the duration of my 4 PM meeting?*
+- Find a MeetingRoom
```
Find a Meeting Room
